### PR TITLE
3P Media: Add author attribution to the Media Element

### LINKED ISSUE #3500: Add author attribution for Unsplash when a media element is focused or hovered
As per Unsplash requirements. See PRD.

### PR DESCRIPTION
**Pull request by winterDroid**
## Summary

Add the author name, ellisized if to long, to the bottom of the media element. 
Just show on hover / focus.
If the actual name container is focused, it gets highlighted.

## Relevant Technical Choices

Hide video duration on hover / focus.

## User-facing changes

ezgif-7-1751cd8627aa

---

Addresses #3500

### CODE REVIEW

File: assets/src/edit-story/components/library/panes/media/common/attribution.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * External dependencies
+ */
+import styled from 'styled-components';
+import PropTypes from 'prop-types';
+import { rgba } from 'polished';
+import { useCallback, useState } from 'react';
+
+const Container = styled.a`
```
Comment (diegovar) [inline assets/src/edit-story/components/library/panes/media/common/attribution.js]: Maybe rename this to Anchor? Container tends to be a div or span.

File: assets/src/edit-story/components/library/panes/media/common/attribution.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  display: block;
+  position: absolute;
+  left: 0;
+  bottom: 0;
+  width: calc(100%);
```
Comment (diegovar) [inline assets/src/edit-story/components/library/panes/media/common/attribution.js]: Can't this just be `width: 100%`?

File: assets/src/edit-story/components/library/panes/media/common/attribution.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  padding: 8px;
+  color: ${({ theme, active }) => rgba(theme.colors.fg.v1, active ? 1.0 : 0.6)};
+  background-color: ${({ theme, active }) =>
+    rgba(theme.colors.bg.v1, active ? 0.8 : 0.6)};
+  font-family: ${({ theme }) => theme.fonts.duration.family};
+  font-size: ${({ theme }) => theme.fonts.duration.size};
+  line-height: ${({ theme }) => theme.fonts.duration.lineHeight};
+  letter-spacing: ${({ theme }) => theme.fonts.duration.letterSpacing};
+  text-overflow: ellipsis;
+  word-break: break-word;
```
Comment (diegovar) [inline assets/src/edit-story/components/library/panes/media/common/attribution.js]: Will this ever happen given we're ellipsizing?

File: assets/src/edit-story/components/library/panes/media/common/attribution.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  overflow: hidden;
+  white-space: nowrap;
+  text-decoration: none;
+`;
+
+const Attribution = ({ author, url }) => {
+  const [active, setActive] = useState(false);
+  const makeActive = useCallback(() => setActive(true), []);
```
Comment (diegovar) [inline assets/src/edit-story/components/library/panes/media/common/attribution.js]: What's the point of using useCallback without deps? You can probably just do `{() => setActive(true)}` below.

File: assets/src/edit-story/components/library/panes/media/common/attribution.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  return (
+    <Container
+      title={author}
+      active={active}
+      onPointerEnter={makeActive}
+      onFocus={makeActive}
+      onPointerLeave={makeInactive}
+      onBlur={makeInactive}
+      href={url}
+      target="_blank"
```
Comment (diegovar) [inline assets/src/edit-story/components/library/panes/media/common/attribution.js:59]: Didn't you get a linter error to add rel={'noreferrer'}? I think you need that.

File: assets/src/edit-story/components/library/panes/media/common/mediaElement.js [hunk 6]
```diff
@@ -250,6 +251,9 @@ const MediaElement = ({
     onInsert(resource, width, height);
   };
 
+  const showAttribution =
+    active && resource.attribution && resource.attribution.author;
```
Comment (diegovar) [inline assets/src/edit-story/components/library/panes/media/common/mediaElement.js]: [nit] You can probably shorthand this to
`active && resource.attribution?.author`

File: assets/src/edit-story/components/library/panes/media/common/mediaElement.js [hunk 7]
```diff
@@ -258,7 +262,8 @@ const MediaElement = ({
     width,
     height,
     onClick,
-    showVideoDetail,
+    showVideoDetail: showVideoDetail && !showAttribution,
```
Comment (diegovar) [inline assets/src/edit-story/components/library/panes/media/common/mediaElement.js]: maybe you can roll up the extra logic into the useEffect where setShowVideoDetail is called? That way showVideoDetail truly means what its name implies.

File: assets/src/edit-story/components/library/panes/media/common/mediaElement.js [hunk 8]
```diff
@@ -381,6 +395,12 @@ function getInnerElement(
         gallery as there's no event when a video's poster loads. */}
         <HiddenPosterImage src={poster} onLoad={makeImageVisible} />
         {showVideoDetail && <Duration>{lengthFormatted}</Duration>}
+        {showAttribution && (
+          <Attribution
+            author={resource.attribution.author.displayName}
+            url={resource.attribution.author.url}
+          />
+        )}
```
Comment (diegovar) [inline assets/src/edit-story/components/library/panes/media/common/mediaElement.js]: Given the attribution is the same for videos and images, can we bring this outside `getInnerElement`?

File: assets/src/edit-story/components/library/panes/media/common/attribution.js (hunk 5 above)
Comment (swissspidy) [reply assets/src/edit-story/components/library/panes/media/common/attribution.js:59]: Yep, needs `rel="noreferrer"`

The linter only recognizes `<a>` and a handful of custom components. "Container" is not one of them. "Link" would be better.

Another reason for renaming :-)

File: assets/src/edit-story/components/library/panes/media/common/attribution.js (hunk 1 above)
Comment (winterDroid) [reply assets/src/edit-story/components/library/panes/media/common/attribution.js]: Renamed to `Link` as suggested by @swissspidy

File: assets/src/edit-story/components/library/panes/media/common/attribution.js (hunk 4 above)
Comment (winterDroid) [reply assets/src/edit-story/components/library/panes/media/common/attribution.js]: It is done the same in the media element itself. Thus I followed the approach ^^ changed.

File: assets/src/edit-story/components/library/panes/media/common/attribution.js (hunk 5 above)
Comment (winterDroid) [reply assets/src/edit-story/components/library/panes/media/common/attribution.js:59]: Added and renamed.

File: assets/src/edit-story/components/library/panes/media/common/mediaElement.js (hunk 6 above)
Comment (winterDroid) [reply assets/src/edit-story/components/library/panes/media/common/mediaElement.js]: Done.

File: assets/src/edit-story/components/library/panes/media/common/mediaElement.js (hunk 8 above)
Comment (winterDroid) [reply assets/src/edit-story/components/library/panes/media/common/mediaElement.js]: Done.

File: assets/src/edit-story/components/library/panes/media/common/attribution.js (hunk 3 above)
Comment (winterDroid) [reply assets/src/edit-story/components/library/panes/media/common/attribution.js]: Removed.

File: assets/src/edit-story/components/library/panes/media/common/attribution.js (hunk 2 above)
Comment (winterDroid) [reply assets/src/edit-story/components/library/panes/media/common/attribution.js]: Done. Somehow in a different setup it was not working without it.

File: assets/src/edit-story/components/library/panes/media/common/mediaElement.js (hunk 7 above)
Comment (winterDroid) [reply assets/src/edit-story/components/library/panes/media/common/mediaElement.js]: Done. But I created a separate setShowVideoDuration and a separate effect.

**Review by diegovar — APPROVED** [review]

**Pull request merged**